cxx/emissions/simple_string.hh

```diff
@@ -0,0 +1,191 @@
+#pragma once
+
+#include <cstdio>
+#include <unordered_map>
+#include "emissions/base.hh"
+#include "distributions/beta_bernoulli.hh"
+
+// A simple string emission model that handles substitutions, insertions
+// and deletions of characters, all without context.  (I.e., their
+// probability doesn't depend on any previous character or the character
+// being substituted, inserted or deleted.)
+
+class SimpleStringEmission : public Emission<std::string> {
+ public:
+  BetaBernoulli substitution;
+  BetaBernoulli insertion;
+  BetaBernoulli deletion;
+
+  SimpleStringEmission() : substitution(nullptr),
+                           insertion(nullptr),
+                           deletion(nullptr) {};
+
+  void incorporate(const std::pair<std::string, std::string>& x) {
+    ++N;
+    corporate(x.first, x.second, true);
+  }
+
+  void unincorporate(const std::pair<std::string, std::string>& x) {
+    --N;
+    corporate(x.first, x.second, false);
+  }
+
+  void corporate(const std::string& clean, const std::string& dirty, bool d) {
+    if (clean.empty()) {
+      // All of dirty must be insertions.
+      for (size_t i = 0; i < dirty.length(); ++i) {
+        d ? insertion.incorporate(1) : insertion.unincorporate(1);
+      }
+      return;
+    }
+
+    if (dirty.empty()) {
+      // All of clean must have be deleted.
+      for (size_t i = 0; i < dirty.length(); ++i) {
+        d ? deletion.incorporate(1) : deletion.unincorporate(1);
+      }
+      return;
+    }
+
+    if (clean[0] == dirty[0]) {
```

Comment:
This is clever and it took me a little while to understand. Consider adding some explanatory comments to the first couple blocks like "Check equality of string prefixes" before line 50 and "Check equality of string suffixes" before line 66.

Reply:
Done.